src/Npgsql/NpgsqlParameter`.cs

```diff
@@ -89,7 +89,7 @@ private protected override void BindCore(bool allowNullReference = false)
         // If we're object typed we should support DBNull, call into base BindCore.
         if (typeof(T) == typeof(object) || TypeInfo!.IsBoxing || _useSubStream)
         {
-            base.BindCore(TypeInfo!.IsBoxing || _useSubStream || allowNullReference);
+            base.BindCore(typeof(T) != typeof(object) && (TypeInfo!.IsBoxing || _useSubStream || allowNullReference));
```

Comment:
This was a latent bug btw, allowNullReference should always be false when we're doing object typing.